[2단계 - 로또(수동)] 포키(권예진) 미션 제출합니다.

YJGwon opened the pull request:
안녕하세요 제이! 포키입니다.
1단계에서 만든 구조를 바탕으로 크게 변경 없이 수동 로또 기능을 추가할 수 있었어요.
솔직히 저는 2단계 기능 추가하기 수월해서 지금의 구조가 마음에 들었는데, 제이의 시각에서 어떻게 보일지 궁금하네요.
지난번에 말씀해주신 캐싱은 후보 번호에 대한 `LottoNumbers`들을 갖고 있으면서 추첨을 담당하는 `LottoWheel`이라는 객체로 구현해봤습니다.

---
질문 드리고 싶은 것이 두 가지 있어요!

- util 형태로 사용되는 Controller에서 static member를 가지는 것, 괜찮을까요?
마지막에 Controller의 method들을 분리하면서 lottos를 member field로 선언한 후 분리했습니다. 전역 변수는 최대한 쓰지 않으려 하고 있지만 lottos는 controller 전역에 걸쳐 많이 사용되는 변수라서 괜찮겠다는 생각이었는데요, 제이는 어떻게 보시나요?

- 원시값은 어디까지 포장해야 할까요?
이번 요구사항 중 `모든 원시 값과 문자열을 포장한다` 라는 항목이 있는데, 그렇다면  `LottoNumbers.from(List<String> inputs)` 처럼 input 또는 output과 관련해서 생기는 문자열도 모두 포장했어야 할까요? 저는 포장해도 딱히 가지고 갈 비즈니스 로직이 없다고 생각해서 포장하지 않았습니다.

YJGwon committed:
docs: 기능 목록에 수동 기능 추가, 도메인 목록 수정

YJGwon committed:
refactor: 랜덤 번호 추첨기 LottoWheel 추가

- LottoNumber로 부터 로또 번호 6개 추첨 책임 분리
- LottoNumber.from()에서 미리 생성해둔 LottoWheel 내의 LottoNumber 반환

YJGwon committed:
refactor: test에서만 쓰이는 LottoNumbers.getSize(), LottoNumbers.get() 삭제

YJGwon committed:
refactor: MatchResult의 member field 모두 final로 변경

YJGwon committed:
fix: LottoWheel.get()에서 index가 아닌 값으로 LottoNumber 검색

YJGwon committed:
feat: 수동 로또 제외한 수량만큼 자동 로또 생성, 출력

수익률 연산 오류 수정 필요

YJGwon committed:
fix: Prize.amount long으로 변경

YJGwon committed:
refactor: PrizeInformationDTO 하나 만드는 from method 분리

YJGwon committed:
feat: 구매할 수동 로또 개수 입력

YJGwon committed:
feat: 수동 로또 번호 입력

YJGwon committed:
feat: 구매한 로또 수량 출력

YJGwon committed:
refactor: Controller에 static member field lottos 추가

lottos는 Controller 전반에 넓게 쓰이기 때문에, 메서드 분리를 위해 member field로 선언

jjj0611 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -32,8 +32,8 @@ public boolean isUnit(int unit) {
         return amount % unit == 0;
     }
```
포키 여기서 Money가 unit을 받는 이유가 뭘까요?
로또 한 장의 가격인 1000원이 로또에 있는데, 아마 이렇게 생각하셨을 것 같아요.
1000원은 로또 한 장의 가격이니까 로또에 있어야 하는거 아닌가? 하고 생각하셨을 것 같은데,
사실 로또 1장의 가격은 정책일 뿐이지 로또와 직접 관련은 없지 않을까요?
1000원으로 1장의 로또를 살 수 있는거 뿐이지, 1000원을 알아야만 로또가 성립하는건 아니라고 생각해서요 :)
포키의 생각은 어떤까요??

YJGwon replied:
1000원이라는 값은 로또의 속성인 것이 더 자연스럽다고 생각한 것 맞습니다! 제가 만약 이 프로그램 내의 로또 가격을 수정하라는 지시를 받으면, Lotto class를 가장 먼저 열어볼 것 같거든요. Lotto 밖으로 꺼내온다면 Money에서 가격을 상수로 갖고 있게 되는 걸까요?

jjj0611 replied:
엇 그건 구현마다 달라질 것 같은데, 지금 구조에선 그럴 확률이 가장 높아보이네요!
금액에 대한 의존이 로또에서 일어나야 하는가?에 대한 고민이 있을 것 같아요.
로또 가격을 수정하라는 것도 맞지만, 좀 더 정확하게는 구매 가격을 수정하라는 것 아닐까요??

포키 혹시 현재 클래스 구조에서의 관계도를 한 번 정리해봐주실 수 있으실까요?
전체 애플리케이션의 흐름과 그 사이에서의 객체들의 역할을 글로 정리해서 저에게 한 번 설명해주세요!

YJGwon replied:
이 것은 작성한 후에 리뷰 요청 드리며 추가한 아래쪽 코멘트에 달아두겠습니다!

jjj0611 commented on README.md:
````diff
@@ -30,30 +30,47 @@
+- [x]  구매할 수동 로또 개수를 입력받는다
+
+```
+수동으로 구매할 로또 수를 입력해 주세요.
+3
+```
+
+- [x]  수동 로또 번호를 입력받는다
+
+```
+수동으로 구매할 번호를 입력해 주세요.
+8, 21, 23, 41, 42, 43
+3, 5, 11, 16, 32, 38
+7, 11, 16, 35, 36, 44
+```
````
수동에 대한 기능도 요구사항부터 잘 정리해주셨네요 👍

jjj0611 commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -10,20 +11,47 @@
+    private final List<Lotto> lottos;
+    private final int count;
+    private final int manualCount;
+    private int purchasedCount;
```
이 객체를 사용자가 구매한 로또라고 했을 때, 위 값들을 들고 있어야 할 필요가 있을까요?
Lotto라는 객체에서 해당 로또가 자동인지 수동인지 정도는 있으면 좋겠지만, 몇장을 자동으로 샀는지, 수동으로 샀는지 필요한 이유가 뭘까요?

반대로 사용자가 구매한 이 로또라는 친구가 생성되는 시점에 몇 장을 살건지, 얼마나 살건지가 결정되는 이유는 (자동, 수동) 로또 번호의 생성 주기가 명확하지 않기 때문이라고 생각하는데, 로또 번호의 생성을 분리해보면 어떨까요?

YJGwon replied:
View에서 보여줘야 해서.. 라기엔 1단계에선 size로 해결했었네요. LottoNumbers의 생성이 Lotto 안에서 이루어지도록 리팩토링 해보면서 count 값의 존재 이유를 다시 한 번 고민해보겠습니다...!

jjj0611 commented on src/main/java/lotto/model/number/LottoNumbers.java:
```diff
@@ -21,7 +21,7 @@ public LottoNumbers(List<LottoNumber> lottoNumbers) {
 
     public static LottoNumbers ofRandomNumbers() {
-        return new LottoNumbers(LottoNumber.draw(NUMBERS_SIZE));
+        return new LottoNumbers(LottoWheel.draw(NUMBERS_SIZE));
```
lottonumbers에서 lotto wheel을 호출하고 있는데, 이걸 한 번 더  wheel로 감싼 이유가 뭘까요??

YJGwon replied:
wheel로 감쌌다는 말의 의미가 잘 이해되지 않아요😥 만약 해당 로직이 LottoWheel이라는 class에게 분리된 이유를 말씀하시는 거라면, LottoNumbers는 로또 번호 조건에 맞는 6개의 로또 번호만을 의미하는 class이기 때문에 45개의 LottoNumber를 캐싱해두고 섞어서 추첨해주는 역할을 별도 class로 분리했습니다.

YJGwon replied:
*outdated로 뜨는 이유는 method 정렬 순서를 변경했기 때문입니다!

jjj0611 replied:
아 결국 LottoNumbers에서는 LottoWheel에서 나온 결과물로 new를 호출해서 생성만 하고 있는 것으로 보여서요.
LottoWheel에서 LottoNumbers를 리턴하지 않은 이유는 뭘까요?

YJGwon replied:
LottoWheel과 LottoNumbers가 서로 한 번은 호출해야 한다면, LottoWheel이 LottoNumbers를 부르는 것 보다 LottoNumbers가 부르는게 낫다고 생각했어요. 이유는 LottoNumbers는 원래도 다른 객체들과 적극적으로 협력하지만 LottoWheel은 다른 객체에서 접근 할 일이 많지 않기 때문입니다. 예를 들어 자동 로또 번호를 필요로 하는 Lotto가 LottoWheel이라는 새로운 객체는 몰라도 되고, 원래 알고 있던 LottoNumbers에 기능이 추가된 형태로 쓸 수 있다면 더 좋을 것이라 생각했어요.

jjj0611 reviewed (changes requested):
안녕하세요 포키! 리뷰어 제이입니다 :) 

---
먼저 질문 주신 내용부터 답변을 드릴게요!

util 형태로 사용되는 Controller에서 static member를 가지는 것, 괜찮을까요?

controller가 util이 맞는지 다시 한 번 고민해보면 좋을 것 같아요. 1단계 때 리뷰 드리지는 않았지만, static 메서드 사용에 대해서는 [이 글](https://tecoble.techcourse.co.kr/post/2020-07-16-static-method/)을 참고해봐도 좋을 것 같아요. 그리고 지금 코드에서 lottos 같은 것을 static member로 가지는 것은 좋지 않습니다. lottos를 static 변수에서 제거해주세요! 왜 좋지 않은지는 static 키워드의 사용에 대해 고민해보시면 좋겠습니다!


원시값은 어디까지 포장해야 할까요?

`저는 포장해도 딱히 가지고 갈 비즈니스 로직이 없다고 생각해서 포장하지 않았습니다.` 포키의 이 주장이 더 중요해요! 포키가 그렇게 느꼈으면 그런거죠! :) 이상했다면 리뷰어가 리뷰를 남겼을겁니다!! 그리고 이런 의문이 든다면 반란군이 되어서 반대해보셔도 돼요 :)

---

코멘트는 간단하고 빠르게 적용할 수 있는 부분까지만 남겼어요!
한번에 모든 코멘트를 드리는 것보다 조금씩 바꿔보는게 좋을 것 같아요!
적용하는 동안 헷갈리거나 이해가 가지 않는 부분이 있으면 편하게 이야기 해주세요 :)

YJGwon committed:
refactor: 수량에 맞는 로또 생성을 담당하는 LottosBuilder 추가

- Lotto에 boolean auto 추가: 자동으로 생성된 경우 true
- Lottos의 get*Count method들 수정: Lotto.isAuto()를 통해 count

YJGwon committed:
docs: README.md LottosBuilder 설명 추가

YJGwon committed:
docs: 팩토리 메소드 외의 정적 메소드 모두 instance로 변경

여러 번 반복적으로 호출되는 것이 아니라 한 번 호출되고 마는 메소드들이 메모리에 남아있지 않게 static 키워드 삭제

YJGwon committed:
refactor: Lottos 내부에서 lottos가 변경될 일이 없어졌기 때문에 생성자에서 불변 처리

getManualCount()에 임시로 추가했던 peek 삭제

YJGwon commented:
안녕하세요 포키입니다!

말씀주신 '로또 번호의 생성주기'라는 말이 제게는 조금 어려워서 나름 이해를 해봤어요. 처음에는 `Lotto`에 넘겨주는 `LottoNumbers`의 생성이 `Lottos`에 있는게 이상하다는 뜻인가? 라고 생각했었다가, `Lottos`가 생성된 이후에 `Lotto`들이 생성되는 것이 잘못되었다고 느꼈어요. 빈 list를 가진 `Lottos`를 만든 후 add method를 활용한다면 lottos는 변경에 대한 위험에 처하니까요.
그래서 Java의 `StringBuilder`를 떠올리며, `List<Lotto>`를 개수에 맞게 차곡 차곡 만들어주는 `LottosBuilder`를 추가했습니다. 의도하신 방향은 아닐 것 같지만, 제 아이디어를 한 번 넣어봤어요.

그리고 static method에 대해서도 링크 걸어주신 글도 읽어보고 다른 글들도 읽어봤습니다. 그 후 [d6ad68b](https://github.com/woowacourse/java-lotto/pull/424/commits/d6ad68ba4a7642db80286ba19740b25d20c302bd)에 남겨두었듯 `여러 번 반복적으로 호출되는 것이 아니라 한 번 호출되고 마는 메소드들이 메모리에 남아있지 않도록` `Controller`, `InputView`, `ResultView`를 모두 동적으로 사용하게 변경했습니다. `Lotto`의 static method인 `countAvailableTickets()` 역시 Money가 로또 가격을 알게 된다면 없앨 수 있었겠지만, 음...이건 아직은 어색하게 느껴져서 일단은 그대로 두었어요.

조금씩 바꿔보자고 말씀해 주셔서, 이번엔 빠르게 완성해봤습니다. 잘 부탁드립니다!
comment로 주신 질문들에 대한 답변은 모두 해당 comment에 달아두었는데, 한 곳에 모으는 것이 더 읽기 편하시다면 살짝 귀띔해주세요 :)

---

### 전체 애플리케이션의 흐름과 그 사이에서의 객체들의 역할
##### 구매 금액과 수량
프로그램이 시작되면, 사용자에게 구매 금액을 입력받아 `Money` 객체를 생성합니다.
사용자가 수동 로또를 몇 장 구매할 지  입력하면, `LottosBuilder`는 `Money`와 수동 로또 장수를 받아 몇 장의 로또를 살 수 있는지 구합니다. 총 구매 수량은 가격을 알고 있는 `Lotto`에게 물어봅니다. `Lotto`는 `Money`의 총액을 모르기 때문에 `Money`에게 자신의 가격을 알려주고 구매 가능한 수량을 구합니다. 
##### 로또 생성
수량이 구해지면, `Controller`는 `LottosBuilder`가 허락할 때 까지 사용자가 입력한 값으로 수동 로또를 생성합니다. 입력값은 숫자별로 쪼개어져 각각 `LottoWheel`이 미리 가지고 있던 `LottoNumber`객체로 변환되고, 이 6개의 숫자는 `LottoNumbers`로 포장되어 Lotto의 값이 됩니다.
남은 수량은 `LottosBuilder`에서 모두 자동 로또로 만들어줍니다. `Lotto`는 `LottoNumbers`에게 랜덤한 번호들을 요청하고, `LottoNumbers`는 `LottoWheel`이 추첨해준 번호들을 `LottoNumbers`로 만들어서 `Lotto`에게 return합니다.
수량만큼 로또가 다 만들어지면, `LottosBuilder`는 이를 `Lottos`로 포장해서 전달합니다.
##### 당첨 번호와 보너스 볼
당첨 번호는 수동 로또와 같은 과정을 거쳐 `LottoNumbers`가 되고, 보너스 볼은 `LottoNumber`가 됩니다. 이 둘은 `WinningLotto`로 합쳐집니다.
##### 당첨 결과
당첨 여부를 알아보기 위해, 먼저 `LottoNumbers`타입인 당첨 번호와 구매한 로또 번호가 서로를 비교해서 겹치는 번호의 수를 구합니다.  `LottoNumbers`에게 보너스 볼 일치 여부까지 물어본 후 , 한 장의 로또 당첨 여부 판단을 위한 이 정보들을 `MatchResult`에 담습니다. `Lottos`에서 모든 로또의 `MatchResult`를 구한 후 이를 `PrizeInformations`에게 전달합니다.
`PrizeInformation`은 한 순위당 몇 장이 당첨인지 그 정보를 저장합니다. 당첨 순위를 결정하는 기준은 `Prize` 안의 각 순위 스스로가 알고 있습니다. 각 순위에게 '이 결과가 네 결과냐'고 물어보아, 그렇다고 대답하는 횟수를 세어 해당 `Prize`의 `PrizeInformation`을 만듭니다.
##### 당첨 통계
`PrizeInformation`은 `Prize`에게 해당 순위 당첨 수를 알려주며 순위별 총 당첨금을 묻습니다. `PrizeInformations`는 각 `PrizeInformation`이 알아온 순위별 당첨금을 합쳐 총 당첨금을 구하고, `Money`에게 가진 금액에 비해 총 당첨금이 얼마나 큰지 물어봅니다. 이 값이 곧 수익률입니다.

jjj0611 reviewed (changes requested):
안녕하세요 포키! 리뷰어 제이입니다.

---

객체간의 관계에 대한 설명 중

 `총 구매 수량은 가격을 알고 있는 Lotto에게 물어봅니다. Lotto는 Money의 총액을 모르기 때문에 Money에게 자신의 가격을 알려주고 구매 가능한 수량을 구합니다.` 

 -> money는 총 구매 수량을 알기 위해 lotto가 필요하고 lotto는 총액을 위해 money가 필요합니다. 결국 총액과 단위 금액이 있으면 해결될 문제이고, 이것이 money 안으로 들어가야하는 것 아닐까요? lotto에 단위 금액이 있어야만 하는 이유는 무엇인가요??


`Lotto는 LottoNumbers에게 랜덤한 번호들을 요청하고, LottoNumbers는 LottoWheel이 추첨해준 번호들을 LottoNumbers로 만들어서 Lotto에게 return합니다. 수량만큼 로또가 다 만들어지면, LottosBuilder는 이를 Lottos로 포장해서 전달합니다.`

-> Lotto라는 이름이 생성을 담당하는것에 적합한 책임일까요...? LottoNumbers라는 친구는 6개의 로또 번호의 묶음이라는 설명이 있는데 여기서 직접 LottoWheel을 요청하는게 적합한 책임일까요? Lotto의 클래스 설명을 보면 로또 한 장을 의미한다고 했는데 LottoNumbers를 보면 6개의 로또 번호의 묶음을 의미한다고 하네요. 로또 한 장과 6개의 로또 번호의 묶음의 차이는 어떤걸까요? 6개의 로또 번호의 묶음이 로또 한 장 아닐까요?

객체 간의 설명 중에 조금 궁금한 부분에 대해서 다시 질문 드려요 :)

---

LottosBuiler를 사용하신 부분은 좀 기발하다고 생각하지만, 자연스럽게 읽히지는 않는 것 같아요.
자동 생성을 담당하는 친구와 수동 생성을 담당하는 친구를 만들어서 자동과 수동 요청에 대한 값으로 생성만 해주면 되지 않을까요??

---

코드 하나하나에 코멘트를 드리는 것보다 전체적인 방향성에 대한 논의를 하는게 더 맞을 것 같아요!
이 코멘트들에 대해서 포키가 이해하는 방향 혹은 느끼는 것들을 자유롭게 이야기 해주시면 논의에 더 좋을 것 같아요 :)

YJGwon committed:
refactor: 로또 한 장의 가격 Money로 이동

로또의 가격은 구매 금액에 관한 처리에 필요한 정보이기 때문

YJGwon committed:
refactor: 중복되는 로직을 가진 Lotto와 LottoNumbers 통합

로또 번호가 6개 있으면 그 자체로 Lotto라고 볼 수 있고, 실제로 Lotto에서 특별한 로직이 추가되지 않기 때문에 통합

YJGwon committed:
docs: 도메인 설계 수정 (LottoNumbers 삭제)

YJGwon committed:
refactor: LottoWheel이 자신이 가지고 있는 LottoNumber를 직접 반환하고, Lotto도 직접 만들어서 반환

- Lotto의 size 정보는 삭제되어야 함
- Lotto를 생성하는 역할은 분리 필요

YJGwon committed:
refactor: LottoWheel -> LottoNumberCache로 변경

- LottoNumberCache: 캐싱해둔 LottoNumber를 꺼내주는 역할
- 로또 생성은 Lotto 클래스 내부에서 처리

YJGwon committed:
refactor: 로또 구매 수량을 구해두는 것이 아니라 로또 생성할 때 마다 money에서 차감

YJGwon committed:
refactor: LottosBuilder -> LottoCart로 rename

Builder 패턴이라 오해할 소지가 있어, 로또를 구매해서 Lottos로 만드는 중간 과정의 의미를 담아 LottoCart로 이름 변경

YJGwon commented:
안녕하세요! 포키입니다.
먼저 로또 생성 횟수 관리에 대한 로직을 거의 변경하지 못했다는 점 부터 말씀드려요...
수동 로또를 생성할 때 마다 `InputView`를 호출해야 하다 보니 `Controller`밖에서 이 횟수를 관리하는 다른 좋은 방법이 떠오르지 않아요 😭
`Controller`에서 바로 횟수만큼 for문을 돌리는 것은 비즈니스로직을 가지는 것과 다름이 없으니 절대 하고싶지 않은데, 그러려면 수동로또 수량을 가진 누군가가 `Controller`가 `InputView.askManualNumbers()`를 호출할 때 마다 그 수량을 차감하고 0이 되면 알려주는 방식말고는 없지 않나... 다른 방법이 있는건가... 라는 생각입니다.

대신, 아래 사항들은 제 생각을 녹여서 리팩토링했습니다.
- 로또 가격의 행방 (Money로 이동했습니다)
- Lotto와 LottoNumbers의 불필요한 분리 (통합했습니다)
- LottoWheel의 요상한 정체성 (LottoNumber들의 캐시 역할임을 명시했습니다)

말씀 주신 과제를 다 해결하지 못하고 리뷰를 요청하는게 조금 찝찝하지만 변경사항에 대한 리뷰를 받으며 힌트를 좀 얻을수 있을까 해서 요청드려요. 이럴 경우엔 요청 이전에 DM을 드렸어야 했던 걸까요? 그렇다면 지적 부탁드립니다.

---
(+) 요청 후 EnumMap이라는 것을 알게 되어서 `PrizeInformation`을 `Prize`를 key로 가지는 EnumMap으로 수정했습니다.

YJGwon committed:
refactor: 당첨 정보 EnumMap<Prize,Integer>로 변경

- 당첨 정보 DTO에서 PrizeDTO 분리
- 수익률 계산 오류 해결 필요

YJGwon committed:
fix: amount가 변경되어 수익률 오류나는 문제 해결

YJGwon committed:
docs: README의 PrizeInformation 항목 PrizeCountMap으로 수정

YJGwon committed:
refactor: 지불 금액 연산식을 복합 대입 연산자로 변경

jjj0611 commented on src/main/java/lotto/model/Money.java:
```diff
@@ -6,12 +6,21 @@
+    private static final int PRICE_LOTTO = 1000;
 
     private final int amount;
+    private int purchasedAmount;
```
돈이라는 객체가 구매된 양을 알고 있네요.
돈의 역할이 맞을까요??

YJGwon replied:
방식 자체보다는 누구의 책임인지가 관건이군요! 잘 분리해보겠습니다.

jjj0611 commented on src/main/java/lotto/model/number/LottoNumberCache.java:
```diff
@@ -0,0 +1,35 @@
+    public static List<LottoNumber> getRandomNumbers(int size) {
+        Collections.shuffle(LOTTO_BALLS);
+        return new ArrayList<>(LOTTO_BALLS.subList(0, size));
+    }
```
이 친구는 cache라고는 했지만, 어떻게 보면 로또번호를 생성해주는 친구에 가까운 것 같아요!

YJGwon replied:
네이밍을 할 때 겉으로 보이는 역할이 아닌 내부 로직을 너무 많이 신경쓰는 것도 문제가 될 수 있겠네요. 바꿔볼게요!

jjj0611 commented on src/main/java/lotto/model/number/LottoNumber.java:
```diff
@@ -1,51 +1,36 @@
-    public static List<LottoNumber> draw(int size) {
-        Collections.shuffle(NOMINEE_NUMBERS);
-        return NOMINEE_NUMBERS.subList(0, size);
+        return allNumbers;
```
이 친구가 public으로 열려 있는한 모든 LottoNumber가 항상 Cache 데이터는 아니겠네요!

jjj0611 commented on src/main/java/lotto/model/prize/Prize.java:
```diff
@@ -40,15 +40,15 @@ private static Prize getPrizeByCount(MatchResult matchResult) {
                 .orElse(NONE);
```
26번 라인부터 40번 라인까지의 로직을 좀 더 단순화할 순 없을까요??

YJGwon replied:
이 부분은 오늘 강의 중 다른 크루가 이야기하는 것을 보고 배운게 있어서 한번 활용해보려고 해요! 깔끔하게 만들어보겠습니다!

jjj0611 commented on src/main/java/lotto/model/prize/PrizeCountMap.java:
```diff
@@ -0,0 +1,45 @@
+        EnumMap<Prize, Integer> prizeMap = new EnumMap<>(Prize.class);
+        Arrays.stream(Prize.values())
+                .filter(prize -> prize != Prize.NONE)
+                .forEach(prize -> prizeMap.put(prize, count(matchResults, prize)));
```
Stream을 사용하셨으니 좀 더 간단하게 표현하기 위해
collect와 Collectors의 기능을 한 번 찾아보면 어떨까요??

jjj0611 commented on src/main/java/lotto/model/prize/MatchResult.java:
```diff
@@ -4,8 +4,8 @@
 public class MatchResult {
-    private int count;
-    private boolean bonus;
+    private final int count;
+    private final boolean bonus;
```
match result는 좋아요!

jjj0611 reviewed (changes requested):
안녕하세요 포키!

말씀드렸던 부분을 잘 이해해주시고 반영해주셨네요!
고민으로 남겨주신 부분에 대해서는 Money쪽 코멘트를 한 번 참고해보시면 좋을 것 같아요!

조금 더 수정되면 좋을 부분들 코멘트 드렸어요! 한 번 고민해보시고 적용해주세요! 

전체적으로 코드가 정리되어지고 있다는 느낌이 많이 드네요 👍
조금만 더 노력해봐요!

아 그리고 DM 관련해서는 편하게 이야기 해주세요!
만약 이해가 잘 가지 않는다면 DM으로 이야기를 좀 더 하고 한 번에 적용하여 리뷰 주시면, 
리뷰를 끊어가지 않아도 되니 편한 부분은 있을 것 같아요!
편하게 요청 주시면 됩니다 :)

YJGwon committed:
refactor: LottoNumber.number -> value로 수정

Class명에서 이미 number라고 명시하고 있어 네이밍 수정
일급 컬렉션 class들은 value라는 네이밍이 헷갈릴 수 있을 것 같아서 보류

YJGwon committed:
refactor: 구매 수량을 관리하는 PurchaseCount 추가

YJGwon committed:
refactor: LottoNumberCache -> LottoNumberFactory로 rename, LottoNumber의 생성 관련 method들 접근제한 default로 변경

YJGwon committed:
refactor: Prize에 Predicate 구현체 winningCondition 추가

YJGwon committed:
refactor(PrizeCountMap): matchResults를 EnumMap으로 만들 때 Collectors.groupingBy() 이용

- Collectors.counting()의 return type에 맞춰 PrizeCountMap의 value도 Long으로 변경

YJGwon committed:
refactor: Money에 필요없는 method 정리

YJGwon committed:
refactor: 사용자 입력값 받는 팩토리 메서드의 parameter 이름 raw<데이터이름> 형식으로 rename

input이라는 변수명은 가리키는 바가 모호하여 변경

YJGwon committed:
feat(PurchaseCount): 음수 입력 예외 처리 추가

PurchaseCountTest.count_10_manual_3() 오류 수정 (기대값 True -> False)

YJGwon commented:
안녕하세요 제이! 포키입니다.
잘 이해했다고 해주시니, 마음이 조금은 놓이네요. 감사합니다 😊 

##### 주요 변화
- 구매 수량을 따로 포장해서 관리하도록 `PurchaseCount`를 추가했습니다.
- 그 동안 접근 제한자는 public 아니면 private이었던 과거를 반성하며, `LottoNumber`의 생성 관련 method 모두 default로 설정했습니다.
- Prize 검사 로직을 `Predicate<MatchResult>`으로 처리했습니다.
- EnumMap 생성 시 `Collectors.groupingBy()`를 이용했습니다.
    - toMap()이 아닌 이유는, toMap()으로 EnumMap을 뽑으려면 키값의 중복을 처리하는 `BinaryOperator`를 작성해야 하는데, EnumMap은 키 값이 중복되지 않을 뿐 더러 데이터 특성 상 나와야 할 키 값은 정해져있기 때문에 그룹화가 더 적절하다고 판단했어요.

그 외에 제가 보기 조금 불편하게 느껴지는 부분들 (주로 naming) 함께 수정해봤어요.
제이가 말해주지 않는 부분들도 스스로 보고 고칠 수 있으면 좋을텐데, 아직은 그런 시야는 제게 없나봐요...😅
하고 싶은 말씀 많으셨을텐데도 꾹 참고 답답한 제가 여기까지 오도록 기다려주셔서 너무 감사해요!
조금만 더 부탁드리겠습니다...! 제이도 바쁘실텐데 화이팅하셔요ㅜㅜ

YJGwon committed:
refactor(PurchaseCount): check method들 validate로 묶음

jjj0611 commented on src/main/java/lotto/Controller.java:
```diff
@@ -1,49 +1,79 @@
-    private static LottoNumbers askWinningNumbers() {
-        String[] winningNumbersInput = InputView.askWinningNumbers();
+    private void purchaseManualLotto(LottoCart lottoCart) {
+        while (lottoCart.isManualAvailable()) {
+            lottoCart.addManualLotto(List.of(inputView.readNumbers()));
+        }
```
isManualAvailable인 상태 동안에
안쪽에서 또 isManualAvailable을 내부적으로 호출해야 하네요!
방어로직으로 만들어주신 것은 이해합니다만,
이 부분에 대한 설계가 바뀌면 이렇게 중복으로 체크하지 않아도 될 것 같아요!

jjj0611 commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -1,41 +1,81 @@
 
-    private final LottoNumbers numbers;
+    private final List<LottoNumber> numbers;
+    private final boolean auto;
```
요건 현재 쓰이는 부분은 없는거죠??

YJGwon replied:
원래는 PurchaseCount의 상태 값이 변경되어 구매 수량을 새로 구해야 했고, 이를 생성된 Lotto 객체의 개수를 세어서 구했기 때문에 자동/수동 판별이 필요했어요. 그런데 PurchaseCount의 상태가 불변이 되면서 필요없어져서 삭제했습니다.

jjj0611 commented on src/main/java/lotto/model/number/LottoNumberFactory.java:
```diff
@@ -0,0 +1,35 @@
+        return LOTTO_BALLS.stream()
+                .filter(lottoNumber -> lottoNumber.getValue() == number)
+                .findFirst()
+                .orElse(new LottoNumber(number));
```
로또 번호를 찾을 때마다 계속 모든 lotto balls를 뒤져야겠네요.
좀 더 좋은 방법 없을까요?? \

jjj0611 commented on src/main/java/lotto/model/prize/Prize.java:
```diff
@@ -1,54 +1,57 @@
-        }
-        return THIRD;
+        this.winningCondition = (matchResult) -> matchResult.isCount(this.matchCount);
     }
```
좋은 방법을 고민해주셨네요!
그런데 생성자에서 호출하는 메서드에서 this.~ 로 할당을 해버리면 가독성이 조금 떨어지는 것 같아요!
좀 더 가독성 있게 바꿔볼까요??

jjj0611 commented on src/main/java/lotto/model/prize/PrizeCountMap.java:
```diff
@@ -0,0 +1,52 @@
+        return Collectors.groupingBy(Prize::getPrize,
+                () -> new EnumMap<>(Prize.class),
+                Collectors.counting());
+    }
```
굳이 메서드로 빼지 않는 것이 가독에 더 좋은 것 같아요!

YJGwon replied:
그러네요! Collectors가 어떤 collector를 데려오는지가 핵심인데 그 부분이 분리되어 있으면 읽기 불편할 수 있겠어요. 반영하겠습니다.

그런데 막상 inline으로 넣어보니까 메서드 호출이 너무 줄줄 되는 느낌이 저는 답답해보여서, method는 분리하되 어떤 collector를 부르는지 명시해봤습니다!

jjj0611 commented on src/main/java/lotto/model/PurchaseCount.java:
```diff
@@ -0,0 +1,58 @@
+
+    public boolean isManualAvailable() {
+        return manual != 0;
+    }
```
이렇게 객체의 상태를 계속 변경하는 것말고 다른 방법은 없을까요?

jjj0611 reviewed (changes requested):
안녕하세요 포키! 리뷰어 제이입니다.

변경한 내용 잘 보았습니다!
재미있게 적용해주신 부분이 있어서 좋네요 :)

고민할만한 부분 코멘트로 남겨 드렸습니다.

더불어서 이제 객체의 역할들이 적절한 이름과 메서드로서 구분이 된다면
객체 위에 있는 주석은 제거해주셔도 좋을 것 같아요!

아직은 주석이 있는게 편한 코드들도 보이는 것 같은데, 
로또 같은 프로그램은 주석 없이도 충분히 설명 가능하답니다!

리팩토링 해주면서 객체의 역할이 잘 분리되었다고 생각하는 친구들은 조금씩 제거해주세요!

YJGwon committed:
refactor(PurchaseCount): 잔여 구매 수량을 생성된 로또 수와 비교하여 확인

- PurchaseCount의 필드 값이 변경되지 않기 때문에 구매 수량 출력에 이용할 수 있음

YJGwon committed:
refactor(Lotto): boolean auto 삭제

수동, 자동 수량을 Lotto 객체를 세어 구하지 않게 되었기 때문에 Lotto의 자동, 수동 여부를 판별하지 않아도 됨

YJGwon committed:
refactor(LottoNumberFactory): 번호 찾을 때 sort한 뒤 index로 검색

YJGwon committed:
refactor(Prize): instanceWinningCondition에 parameter 추가

this 키워드 삭제로 가독성 향상

YJGwon committed:
refactor(PrizeCountMap): get...EnumMapCollector -> groupToEnumMapCollector로 rename

어떻게 collect하는지 알 수 있도록 rename

YJGwon committed:
refactor: 불필요한 주석 제거

YJGwon commented:
안녕하세요 제이~!
이번에 말씀주신 부분은 다 반영해두었는데요, 제가 한 가지 더 시도해보고 싶은게 있어서 여쭤보려 합니다!

다른 크루와 이번 미션에 대해 이야기 하다가, 그 크루가 저에게 **Controller에서 View에 대한 의존성을 제거하고 main method를 통해 서로 요청을 주고받는 방식**을 이야기해주었어요.
- Controller가 마치 library처럼 쓰일 수 있게 되고
- Controller와의 소통도 메세지를 주고받는 방식으로 진행된다는 점
- 그리고 무엇보다 M, V, C 각 계층이 정확히 분리되는 점
에서 저도 장점을 느끼고 시도해보고싶은 생각이 들었습니다.

그런데 저희 프리코스를 떠올려보면 예외 시 재시작 처리를 했었는데, 만약 View를 main에서 호출한다면 재시작 로직이 main에서도 표현되어야 하잖아요.. 이런 재시작 로직이 main에 표현되어도 괜찮은건가 싶은 생각이 들어서 제이에게 이야기해보고 싶었어요. 컨트롤러 리팩토링에 대한 제이의 의견 나누어주시면 감사하겠습니다!

YJGwon committed:
refactor: view class들 멤버 필드 키워드 수정, PurchaseCountDTO.of -> from으로 rename

jjj0611 commented on src/main/java/lotto/model/PurchaseCount.java:
```diff
@@ -0,0 +1,58 @@
+public class PurchaseCount {
+    private static final String ERROR_COUNT_OVER = "[ERROR] 구매할 수 있는 수량을 초과했습니다";
+    private static final String ERROR_NEGATIVE = "[ERROR] 구매 수량은 0보다 작을 수 없습니다";
+    public static final String ERROR_TYPE = "[ERROR] 로또 구매 수량은 숫자로만 입력해주세요";
```
public으로 열려 있네요!

YJGwon replied:
이런걸 놓치는건 정말 부끄럽네요.. 죄송합니다 더 꼼꼼히 확인하겠습니다!

jjj0611 commented on src/main/java/lotto/model/LottoCart.java:
```diff
@@ -0,0 +1,36 @@
+        }
+    }
+
+    public boolean isManualAvailable() {
+        return !purchaseCount.isManual(lottos.size());
+    }
```
수동 구매가 먼저 이뤄지는게 아닌 자동 구매가 먼저 이뤄지면, 이 로직은 실패 하겠네요.

YJGwon replied:
전혀 생각 못했네요! ㅜㅜ Lottos에서도 manual과 auto를 구분해야 할 필요가 있었네요. 삭제했던 Lotto의 boolean auto를 되살려서 더 철저하게 만들어보겠습니다.

jjj0611 commented on src/main/java/lotto/model/LottoCart.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public void addAutoLottos() {
+        while (!purchaseCount.isTotal(lottos.size())) {
+            this.lottos.add(Lotto.ofRandom());
+        }
```
Random하게 생성하는 로직에서 원하는 개수를 입력할 순 없을까요?

YJGwon replied:
이해했습니다! Lotto.ofRandom()에서 parameter로 개수를 받아서 아예 List를 반환할 수 있도록 리팩토링 하겠습니다.

jjj0611 commented on src/main/java/lotto/view/InputView.java:
```diff
@@ -3,29 +3,48 @@
+    public String askMoneyAmount() {
         System.out.println(MESSAGE_MONEY_AMOUNT);
         return scanner.nextLine();
     }
 
-    public static String[] askWinningNumbers() {
+    public String askManualCount() {
+        System.out.println(MESSAGE_MANUAL_COUNT);
+        return scanner.nextLine();
+    }
+
+    public void askManualNumbers() {
+        System.out.println(MESSAGE_MANUAL_NUMBERS);
+    }
+
+    public String[] readNumbers() {
+        return splitByComma(scanner.nextLine());
+    }
+
+    public String[] askWinningNumbers() {
         System.out.println(MESSAGE_WINNING_NUMBERS);
         return splitByComma(scanner.nextLine());
     }
 
-    public static String askBonusNumber() {
+    public String askBonusNumber() {
         System.out.println(MESSAGE_BONUS_NUMBER);
         return scanner.nextLine();
     }
 
-    private static String[] splitByComma(String string) {
+    private String[] splitByComma(String string) {
         return string.split(WINNING_NUMBER_DELIMITER);
     }
```
input view의 모든 메서드가 String을 반환해야 한다는 강박에서 벗어나도 될 것 같아요!
askMoney를 할 때 int가 반환되는 것이 String이 반환되는 것보다 더 직관적일 것 같아요.
view에 대한 로직은 어느정도 자유롭게 설정하셔도 좋지만, String이 반환될 때와 int등 숫자가 반환될 때의 차이점을 한 번 구현해보셔도 좋을 것 같아요 :)

YJGwon replied:
기본적인 validation일지라도 domain에 주는 것이 더 타당할 것만 같은 생각은 강박일지도 모르겠네요. 한 번 구현해보면서 제가 느껴보도록 하겠습니다. view에대한 제약이 제가 생각한 것 보다는 덜 빡빡한것 같기도 하네요..? view에 대한 관점을 많이 찾아봐야겠어요. 감사합니다!

jjj0611 reviewed (approved):
안녕하세요 포키! 리뷰어 제이입니다!

코멘트 드린 부분 잘 반영해주셨는데,
조금 아쉽다고 생각하는 부분에 대해서는 코멘트 남겨두었어요!

로또 미션 동안 피드백을 잘 반영해주셔서 감사합니다!

코멘트 드린거 확인해보시고, 적용도 한 번 해보시면 좋을 것 같아요!

미션 진행하면서 학습하셔야할 부분에 대해서는 다 학습하신 것 같아서 코드는 여기서 머지할게요!

다만, 질문 주셨던 내용 중에 main에 재시작(혹은 재 입력 요청)이 등장해도 되는가에 대한 이야기를 해주셨는데,
포키가 원하는 방향을 제가 잘 이해한게 맞다면, 굳이 main에 등장시키지 않고도 구현할 수 있을 것 같아요.

그리고 현재 포키의 애플리케이션은 재시작 로직이 없는 상태 아닌가요...?
재시작 로직이란 것에 대해서 제가 잘 이해한건지를 모르겠네요 😅

관련해서 추가적으로 더 궁금한 부분 있으면 DM 주세요!
로또 코드는 여기서 머지하겠습니다! 
고생 많으셨어요
